### ISSUE TITLE
With which version of TypeScript is it compatible?

### ISSUE THREAD

**Issue by lcswillems** [swc-project/swc#1161]
Can I compile code written in TypeScript 4.1 with swc?

I can't see any mention about what versions of TypeScript are supported.

**Comment by kitsonk** [comment]
My understanding is that swc isn't coupled to a specific version of TypeScript, like a lot of other transpilers.  There are three types of changes to TypeScript that generally can be made, two of which would impact swc:

- Additional ECMAScript/JavaScript syntax.  TypeScript adds support for Stage 3 ECMAScript proposals, swc adds them as well, usually in advance of them landing in TypeScript (because the release cycle is faster at the moment).  So if a release of TypeScript supports it (even a beta release like 4.1 is at the moment) swc almost certainly does.
- Additional type syntax.  In most cases, the semantics around how to erase types and type information are well understood and really don't change that often, even if TypeScript adds loads of new features.  Most of the time this has no impact on swc.  If there is something, even in a beta release of TypeScript that doesn't get stripped properly, it is a defect and needs to be fixed.
- Additional type checking.  Currently swc doesn't do any type checking, so any changes to how TypeScript checks or enforces types have no impact on swc.

So long story short, any version of TypeScript is supported, and if it doesn't work, it is likely a defect.  The one exception to this rule is that swc (or other transpilers) cannot support the handful of type directed emits that `tsc` supports.  These restrictions are simliar to the Babel ones, documented here: https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats

**Comment by lcswillems** [comment]
Thank @kitsonk for the time you took to explain me this! I understand the reasoning. But I still don't know if I can compile my 4.1 code.

It is a big issue for me that there is no mention of which versions of TypeScript it supports. I (and other people I think) will not take the time to compile their codebase to see if it works with swc. How could people know easily if swc will work with their code?

Furthermore, how do you explain swc is not used everywhere? If it is much faster than babel and tsc and it supports the same functionnalities, it should be a no brainer choice and every should use it. So I guess there is an issue somewhere (maybe in the communication).
